tests/types/test_type_hints.py

```diff
@@ -0,0 +1,325 @@
+import datetime
+from typing import Any, Dict, List, Optional, Union
+
+import pydantic
+import pytest
+
+from mlflow.exceptions import MlflowException
+from mlflow.types.schema import AnyType, Array, ColSpec, DataType, Map, Object, Property, Schema
+from mlflow.types.type_hints import (
+    _infer_schema_from_type_hint,
+    _validate_example_against_type_hint,
+)
+
+
+class CustomModel(pydantic.BaseModel):
+    long_field: int
+    str_field: str
+    bool_field: Optional[bool] = None
```

Comment:
```suggestion
    bool_field: bool
    optional_field: Optional[str] = None,
```

can we split this?

Reply:
Is `<type>| None` supported?

Reply:
> Is <type>| None supported?
Not yet, I think we can support later if there's a strong FR

Reply:
ref: https://peps.python.org/pep-0604/

Reply:
I think we can add support later, we're running our CI in python 3.9, and it's not supported until 3.10. I can add a TODO

Reply:
How does it help?
```
x = int | str
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'type' and 'type'
```

Reply:
`from __futures__ import annotations` enables lazy type hint evaluation, which means "hey python, ignore type annotations". You need to use `int | str` in a type annotation.

Reply:
@serena-ruan LGTM once this is supported.

Reply:
Updated Union type to support |, and validated it works in 3.10
<img width="1061" alt="image" src="https://github.com/user-attachments/assets/cba122d8-9163-4020-a52e-4e9b4b7790ee">

While I cannot directly add test in this PR since `x|y` is not supported in python 3.9 and _infer_schema_from_type_hint directly accepts the type hint. I'll add tests for this in the pyfunc PR after updating _infer_signature_from_type_hints